Notation.Plugin.AzureKeyVault/Certificate/CertificateChain.cs

```diff
@@ -8,19 +9,102 @@ namespace Notation.Plugin.AzureKeyVault.Certificate
-        /// Build a certificate chain from a leaf certificate and a 
-        /// certificate bundle.
+        /// Build the certificate chain from the certificates by matching
+        /// the issuer and subject distinguished name.
         /// 
-        /// Note: the method doen't check the validity of the chain.
-        /// <param name="certificateBundle">The certificate bundle.</param>
-        /// <param name="leafCert">The leaf certificate.</param>
-        /// <returns>A list of raw certificates in a chain.</returns>
+        /// Note: It doesn't verify the certificate path.
         /// </summary>
-        public static List<byte[]> Build(X509Certificate2 leafCert, X509Certificate2Collection certificateBundle)
+        /// <param name="certs"></param>
+        /// <returns>
+        /// The certificate chain. The order is from leaf to root.
+        /// </returns>
+        /// <exception cref="PluginException"></exception>
+        public static X509Certificate2Collection Build(X509Certificate2Collection certs)
         {
-            X509Certificate2Collection chain = new X509Certificate2Collection(leafCert);
-            chain.AddRange(certificateBundle);
-            return chain.Select(x => x.RawData).ToList();
+            /*
+            Valid case:
+            1. self-signed leaf cert
+            2. cert chain: CA1 -> CA2 -> leaf
+
+            Invalid cases:
+            1. empty
+            2. non-self-signed leaf cert
+            3. missing intermediate cert: CA1 -> (x) -> leaf 
+            4. missing root cert: (x) -> CA2 -> leaf
+            5. unnecessary intermediate cert: CA1 -> CA2 -> leaf
+                                              (x) -> CA3
+            6. unnecessary root: CA1 -> CA2 -> leaf
+                                 CA3
+            6. unnecessary cert: CA1 -> CA2 -> leaf
+                                 (x) -> CA3
+            7. duplicated certs: CA1 -> CA2 -> leaf
+                                  └───> CA2
+            8. multiple leaf certs: CA1 -> CA2 -> leaf1
+                                            └───> leaf2
+            9. multple chain: CA1 -> CA2 -> leaf1
+                              CA3 -> leaf2
+            */
+            if (certs.Count == 0)
+            {
+                throw new PluginException("The certificates that need to be built into a chain are empty.");
+            }
+
+            if (certs.Count == 1)
+            {
+                if (certs[0].SubjectName.Name != certs[0].IssuerName.Name)
+                {
+                    throw new PluginException("Only obtained one certificate but it is not a self-signed certificate. Please complete the certificate bundle by `ca_certs` through plugin config.");
+                }
+                return certs;
+            }
+
+            // subject distinguished name -> certificate map
+            var certMap = new Dictionary<string, X509Certificate2>();
+            var issuerSet = new HashSet<string>();
```

Comment:
Will `caSet` be a better name than `issuerSet`?

Reply:
Updated.